rclcpp/include/rclcpp/experimental/buffers/intra_process_buffer.hpp

```diff
@@ -94,6 +95,10 @@ class TypedIntraProcessBuffer : public IntraProcessBuffer<MessageT, Alloc, Messa
 
     buffer_ = std::move(buffer_impl);
 
+    TRACEPOINT(
+      rclcpp_buffer_to_ipb,
+      static_cast<const void *>(buffer_.get()),
+      static_cast<const void *>(this));
```

Comment:
This is a tricky question, but in general we don't have access to everything everywhere ~~all at once~~. For example, we can't get the buffer's capacity here, since `BufferImplementationBase` doesn't expose the capacity. From `RingBufferImplementation`'s constructor (`rclcpp_construct_ring_buffer` tracepoint), we don't have access to the overall buffer. We might be able to trigger `rclcpp_construct_ring_buffer` and `rclcpp_buffer_to_ipb` in `create_intra_process_buffer()` (i.e., link the `BufferImplementationBase` to the `*IntraProcessBuffer`), but in general we try to do this kind of linking at the source of truth, which is the constructor.

I did notice that, technically, we don't really need to record the buffer's capacity, since it's equal to the QoS depth (which the `rcl_subscription_init` tracepoint records): https://github.com/ros2/rclcpp/blob/1688f05aead3968bcea6cab27b346260f633f0b0/rclcpp/include/rclcpp/experimental/create_intra_process_buffer.hpp#L45. But, again, if that ever changes, we won't need to update the instrumentation since we instrumented the constructor directly.

Reply:
@christophebedard thanks for the explanation.

> but in general we try to do this kind of linking at the source of truth, which is the constructor.

this really makes sense to me.